trl/trainer/grpo_trainer.py

```diff
@@ -1413,6 +1434,53 @@ def _generate_and_score_completions(
             sampling_per_token_logps,
             forward_kwargs,
         ) = self._generate(prompts, images)
+        completions = self.processing_class.batch_decode(completion_ids, skip_special_tokens=True)
+        tool_calls = [extract_tool_calls(completion) for completion in completions]
+        tool_results = [self._tool_dict[tc["name"]](**tc["arguments"]) if tc else None for tc in tool_calls]
+        tool_messages = [
+            [{"role": "tool", "name": tc["name"], "content": str(tr)}] if tc else None
+            for tc, tr in zip(tool_calls, tool_results)
+        ]
+        new_prompts = [
+            p + [{"role": "user", "content": c}] + t for p, c, t in zip(prompts, completions, tool_messages) if t
+        ]
+        needs_tool = torch.tensor([tc is not None for tc in tool_calls], device=device)
+        if new_prompts:
+            (
+                new_prompt_ids,
+                new_completion_ids,
+                new_prompt_mask,
+                new_completion_mask,
+                new_num_items_in_batch,
+                new_sampling_per_token_logps,
+                new_forward_kwargs,
+            ) = self._generate(new_prompts, images)
+            num_tool_ids = new_prompt_mask.sum(-1) - torch.cat(
+                [prompt_mask[needs_tool], completion_mask[needs_tool]], dim=1
+            ).sum(-1)
+            tool_ids = [ids[-num:] for ids, num in zip(new_prompt_ids, num_tool_ids)]
+            tool_mask = [torch.ones_like(ids) for ids in tool_ids]
```

Comment:
Would be cool to have a unit test for this masking so we're confident it is behaving as expected